### PR TITLE
Add check that public asset is restricted

### PR DESCRIPTION
**Pull request by ghzlatarev**
## Description

* Add a check that the public asset of a receiver post is not unrestricted, but the default. Otherwise we give a potential attacker control over the nullifier values for no particular reason.
* Currently there's no obvious attack vector but if we end up implementing an exchange where we allow some rules (like swapping the public assets of Utxos) then things may start getting dangerous
* Test cases

---

Before we can approve this PR for merge, please make sure that **all** the following items have been checked off:
- [ ] Connected to an issue with discussion and accepted design using zenhub "Connect issue" button below
- [x] Added **one** label out of the `L-` group to this PR
- [x] Added **one or more** labels from the `A-` and `C-` groups to this PR
- [x] Explicitly labelled `A-calamari` and/or `A-manta` if your changes are meant for/impact either of these (CI depends on it)
- [x] Re-reviewed `Files changed` in the Github PR explorer.

Situational Notes:
- If adding functionality, write unit tests!
- If importing a new pallet, choose a proper module index for it, and allow it in `BaseFilter`. Ensure **every** extrinsic works from front-end. If there's corresponding tool, ensure both work for each other.
- If needed, update our Javascript/Typescript APIs. These APIs are officially used by exchanges or community developers.
- If modifying existing runtime storage items, make sure to implement storage migrations for the runtime and test them with `try-runtime`. This includes migrations inherited from upstream changes, and you can search the diffs for modifications of `#[pallet::storage]` items to check for any.

### CODE REVIEW

File: pallets/manta-pay/src/lib.rs [hunk 1]
```diff
@@ -270,6 +283,15 @@ pub mod pallet {
                     && post.sink_accounts.is_empty(),
                 Error::<T>::InvalidShape
             );
+            for post in post.receiver_posts.clone() {
```
Comment (zqhxuyuan) [inline pallets/manta-pay/src/lib.rs]: is it better to use iter() instead of clone()?

Comment (ghzlatarev) [reply pallets/manta-pay/src/lib.rs]: indeed.

File: pallets/manta-pay/src/lib.rs [hunk 2]
```diff
@@ -236,6 +243,12 @@ pub mod pallet {
                     && post.sink_accounts.len() == 1,
                 Error::<T>::InvalidShape
             );
+            if !post.receiver_posts[0].utxo.is_transparent {
+                ensure!(
+                    post.receiver_posts[0].utxo.public_asset == Asset::default(),
```
Comment (ferrell-code) [inline pallets/manta-pay/src/lib.rs]: Does Asset not implement `zero()`?

**Review by ferrell-code — COMMENTED** [review]
one question looks good

File: pallets/manta-pay/src/lib.rs (hunk 2 above)
Comment (ghzlatarev) [reply pallets/manta-pay/src/lib.rs]: No but yeah i think its better to implement and avoid default()

**Review by ferrell-code — DISMISSED** [review]

**Review by ferrell-code — DISMISSED** [review]

File: .github/workflows/run_all_benchmarks.yml [hunk 3]
```diff
@@ -67,6 +67,7 @@ jobs:
           rustup toolchain install nightly-2023-03-03
           rustup default nightly-2023-03-03
           rustup target add wasm32-unknown-unknown
+          rustup default nightly-2023-01-01
```
Comment (ferrell-code) [inline .github/workflows/run_all_benchmarks.yml]: I think accidental duplicate here

File: .github/workflows/run_linters.yml [hunk 4]
```diff
@@ -78,7 +78,7 @@ jobs:
           curl -s https://sh.rustup.rs -sSf | sh -s -- -y
           source ${HOME}/.cargo/env
           rustup toolchain install nightly-2023-03-13
-          rustup default nightly-2023-03-13
```
Comment (ferrell-code) [inline .github/workflows/run_linters.yml]: Not sure if this change is onpurpose

**Review by ferrell-code — APPROVED** [review]

**Review by Apokalip — APPROVED** [review]

**Pull request merged**